### PR TITLE
LanceDB Integration

### PR DESCRIPTION
**Pull request by AyushExel**
LanceDB is a open-source, serverless, setup-free, multi-modal vector database.

@NivekT , regarding your comments about query_builder - it is essentially a function to query the database which accepts some optional arguments that can be passed with query_args. But feel free to edit of something doesn't make sense

### CODE REVIEW

**Comment by NivekT** [comment]
That makes sense about `query_builder`.

1. If I understand correctly, it is doing an approximate nearest neighbor search, what happens if they want to compare that vs a Python full text search (the experimental feature)? Is that something that we want to support here.
2. Within nearest neighbor search, is that the setup always like it is shown? Or should we allow them to override that default `query_builder`? What if they want to pass in `nprobes`, `refine_factor`, etc? The [Weaviate example](https://github.com/hegelai/prompttools/blob/main/examples/notebooks/WeaviateExperiment.ipynb) has some `query_builder` like that.

File: prompttools/experiment/experiments/lancedb_experiment.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    @staticmethod
+    def _extract_lancedb_dists(output: pd.DataFrame) -> list[tuple[str, float]]:
+        r"""Helper function to get distances between documents from LanceDB."""
+        return output.to_dict(orient="list")["score"]
+
+    @staticmethod
+    def _extract_lancedb_docs(output: pd.DataFrame) -> list[tuple[str, float]]:
+        r"""Helper function to get distances between documents from LanceDB."""
+        return output.to_dict(orient="list")["text"]
```
Comment (NivekT) [inline prompttools/experiment/experiments/lancedb_experiment.py:174]: Any other metrics extracting from the raw LanceDB response?

**Review by NivekT — APPROVED** [review]
Thank you! Have a look at my comment to see if there is anything to add. Otherwise, happy to merge as it is.

**Comment by AyushExel** [comment]
> That makes sense about `query_builder`.
> 
> 1. If I understand correctly, it is doing an approximate nearest neighbor search, what happens if they want to compare that vs a Python full text search (the experimental feature)? Is that something that we want to support here.
> 2. Within nearest neighbor search, is that the setup always like it is shown? Or should we allow them to override that default `query_builder`? What if they want to pass in `nprobes`, `refine_factor`, etc? The [Weaviate example](https://github.com/hegelai/prompttools/blob/main/examples/notebooks/WeaviateExperiment.ipynb) has some `query_builder` like that.

1. I think we can allow FTS but then it would add some setup setps, like installing tantivy and rust. So maybe let's wait till it becomes mature (we're working on it)
2. Yeah we can allow users to scpecify, `n_probes` and `refine_factor` but I think that should still be included in the `query_args` dict, as providing different functions as input and also different args as input might not be needed as we can simply customize the operation by passing different argument parameters?

**Comment by NivekT** [comment]
Thanks for the contribution and answers. We can merge this as it is and expand it based on user requests.

**Pull request merged**